File: DOODL./OnboardingStorage.swift
import Foundation

struct OnboardingStorage {
    private static let defaults = UserDefaults.standard
    private enum Keys {
        static let profileId = "onboarding_profile_id"
        static let username = "onboarding_username"
        static let avatarURL = "onboarding_avatar_url"
        static let pairingCode = "onboarding_pairing_code"
        static let joinedCode = "onboarding_joined_code"
        static let joinedCodes = "onboarding_joined_codes"
    }

    static func save(profileId: String, username: String, avatarURL: URL?, pairingCode: String, joinedCode: String? = nil) {
        defaults.set(profileId, forKey: Keys.profileId)
        defaults.set(username, forKey: Keys.username)
        defaults.set(avatarURL?.absoluteString, forKey: Keys.avatarURL)
        defaults.set(pairingCode, forKey: Keys.pairingCode)
        if let joinedCode {
            defaults.set(joinedCode, forKey: Keys.joinedCode)
        } else {
            defaults.removeObject(forKey: Keys.joinedCode)
        }
    }

    static func load() -> (profileId: String, username: String, avatarURL: URL?, pairingCode: String, joinedCode: String?)? {
        guard
            let profileId = defaults.string(forKey: Keys.profileId),
            let username = defaults.string(forKey: Keys.username),
            let pairingCode = defaults.string(forKey: Keys.pairingCode)
        else {
            return nil
        }
        let avatarURLString = defaults.string(forKey: Keys.avatarURL)
        let avatarURL = avatarURLString.flatMap { URL(string: $0) }
        let joinedCode = defaults.string(forKey: Keys.joinedCode)
        return (profileId, username, avatarURL, pairingCode, joinedCode)
    }

    static func saveJoinedCode(_ code: String) {
        defaults.set(code, forKey: Keys.joinedCode)
    }

    static func saveJoinedCodes(_ codes: [String]) {
        let cleaned = Array(Set(codes.map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }))
            .filter { !$0.isEmpty }
            .sorted()
        if let data = try? JSONEncoder().encode(cleaned) {
            defaults.set(data, forKey: Keys.joinedCodes)
        }
    }

    static func loadJoinedCodes() -> [String] {
        if let data = defaults.data(forKey: Keys.joinedCodes),
           let decoded = (try? JSONDecoder().decode([String].self, from: data)) {
            return Array(Set(decoded.map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }))
                .filter { !$0.isEmpty }
                .sorted()
        }
        // Backward-compatible fallback: if old installs only stored a single joined code, return it.
        if let single = defaults.string(forKey: Keys.joinedCode)?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
           !single.isEmpty {
            return [single]
        }
        return []
    }

    static func saveProfileId(_ id: String) {
        defaults.set(id, forKey: Keys.profileId)
    }

    static func clearJoinedCode() {
        defaults.removeObject(forKey: Keys.joinedCode)
    }

    static func clearJoinedCodes() {
        defaults.removeObject(forKey: Keys.joinedCodes)
    }

    static func clear() {
        defaults.removeObject(forKey: Keys.profileId)
        defaults.removeObject(forKey: Keys.username)
        defaults.removeObject(forKey: Keys.avatarURL)
        defaults.removeObject(forKey: Keys.pairingCode)
        defaults.removeObject(forKey: Keys.joinedCode)
        defaults.removeObject(forKey: Keys.joinedCodes)
    }
}

File: android_flutter/lib/storage/onboarding_storage.dart
import 'package:shared_preferences/shared_preferences.dart';
import '../models/onboarding.dart';
import 'dart:convert';

class OnboardingStorage {
  static const _kProfileId = 'onboarding_profile_id';
  static const _kUsername = 'onboarding_username';
  static const _kPairingCode = 'onboarding_pairing_code';
  static const _kAvatarUrl = 'onboarding_avatar_url';
  static const _kJoinedCode = 'onboarding_joined_code';
  static const _kJoinedCodes = 'onboarding_joined_codes';
  static const _kTutorialShown = 'onboarding_tutorial_shown_v1';

  static Future<OnboardingData?> load() async {
    final prefs = await SharedPreferences.getInstance();
    final profileId = prefs.getString(_kProfileId);
    final username = prefs.getString(_kUsername);
    final pairingCode = prefs.getString(_kPairingCode);
    if (profileId == null || username == null || pairingCode == null)
      return null;

    final joinedCodes = loadJoinedCodesSync(prefs);
    return OnboardingData(
      profileId: profileId,
      username: username,
      pairingCode: pairingCode,
      avatarUrl: prefs.getString(_kAvatarUrl),
      joinedCode: prefs.getString(_kJoinedCode),
      joinedCodes: joinedCodes,
    );
  }

  static Future<void> save(OnboardingData data) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_kProfileId, data.profileId);
    await prefs.setString(_kUsername, data.username);
    await prefs.setString(_kPairingCode, data.pairingCode);
    if (data.avatarUrl != null) {
      await prefs.setString(_kAvatarUrl, data.avatarUrl!);
    } else {
      await prefs.remove(_kAvatarUrl);
    }
    if (data.joinedCode != null) {
      await prefs.setString(_kJoinedCode, data.joinedCode!);
    } else {
      await prefs.remove(_kJoinedCode);
    }
    await saveJoinedCodes(data.joinedCodes);
  }

  static List<String> loadJoinedCodesSync(SharedPreferences prefs) {
    final raw = prefs.getString(_kJoinedCodes);
    if (raw != null && raw.isNotEmpty) {
      final decoded = jsonDecode(raw);
      if (decoded is List) {
        final items = decoded
            .map((e) => e.toString().trim().toLowerCase())
            .where((e) => e.isNotEmpty)
            .toSet();
        return items.toList()..sort();
      }
    }
    final single = prefs.getString(_kJoinedCode)?.trim().toLowerCase();
    if (single != null && single.isNotEmpty) return [single];
    return const [];
  }

  static Future<List<String>> loadJoinedCodes() async {
    final prefs = await SharedPreferences.getInstance();
    return loadJoinedCodesSync(prefs);
  }

  static Future<void> saveJoinedCodes(List<String> codes) async {
    final prefs = await SharedPreferences.getInstance();
    final cleaned = codes
        .map((e) => e.trim().toLowerCase())
        .where((e) => e.isNotEmpty)
        .toSet()
        .toList()
      ..sort();
    await prefs.setString(_kJoinedCodes, jsonEncode(cleaned));
  }

  static Future<void> saveJoinedCode(String? code) async {
    final prefs = await SharedPreferences.getInstance();
    final c = code?.trim().toLowerCase();
    if (c == null || c.isEmpty) {
      await prefs.remove(_kJoinedCode);
    } else {
      await prefs.setString(_kJoinedCode, c);
    }
  }

  static Future<void> addJoinedCode(String code) async {
    final prefs = await SharedPreferences.getInstance();
    final current = loadJoinedCodesSync(prefs);
    final next = {...current, code.trim().toLowerCase()}
        .where((e) => e.isNotEmpty)
        .toList()
      ..sort();
    await prefs.setString(_kJoinedCodes, jsonEncode(next));
    await prefs.setString(_kJoinedCode, code.trim().toLowerCase());
  }

  static Future<void> removeJoinedCode(String code) async {
    final prefs = await SharedPreferences.getInstance();
    final normalized = code.trim().toLowerCase();
    final current = loadJoinedCodesSync(prefs);
    final next = current.where((c) => c != normalized).toList();
    await prefs.setString(_kJoinedCodes, jsonEncode(next));
    final active = prefs.getString(_kJoinedCode)?.trim().toLowerCase();
    if (active == normalized) {
      await prefs.remove(_kJoinedCode);
    }
  }

  static Future<void> clear() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_kProfileId);
    await prefs.remove(_kUsername);
    await prefs.remove(_kPairingCode);
    await prefs.remove(_kAvatarUrl);
    await prefs.remove(_kJoinedCode);
    await prefs.remove(_kJoinedCodes);
    await prefs.remove(_kTutorialShown);
  }

  static Future<bool> isTutorialShown() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_kTutorialShown) ?? false;
  }

  static Future<void> setTutorialShown() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_kTutorialShown, true);
  }
}
